tianshou/policy/modelfree/bcq.py

```diff
@@ -0,0 +1,124 @@
+import torch
+import torch.nn.functional as F
+import numpy as np
+from typing import Any, Dict, Union, Optional
+
+from tianshou.data import Batch
+from tianshou.policy import BasePolicy
+from copy import deepcopy
+
+
+class BCQPolicy(BasePolicy):
+    """Implementation discrete BCQ algorithm. Some code is from
+    https://github.com/sfujim/BCQ/tree/master/discrete_BCQ
+
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        tau: float,
+        target_update_frequency: int,
+        device: str,
+        gamma: float,
+        imitation_logits_penalty: float,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._policy_net = model
+        self._policy_net.to(device)
+        self._optimizer = optim
+        self._cnt = 0
+        self._device = device
+        self._gamma = gamma
+        self._tau = tau
+        self._target_net = deepcopy(self._policy_net)
+        self._target_net.eval()
+        self._target_net.to(device)
+        self._target_update_frequency = target_update_frequency
+        self._imitation_logits_penalty = imitation_logits_penalty
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        **kwargs: Any,
+    ) -> Batch:
+        batch.to_torch()
+
+        state = batch.obs.to(self._device)
+        q, imt, _ = self._policy_net(state.float())
+        imt = imt.exp()
+        imt = (imt / imt.max(1, keepdim=True)[0] > self._tau).float()
+
+        # Use large negative number to mask actions from argmax
+        action = (imt * q + (1.0 - imt) * -1e8).argmax(1)
+
+        return Batch(act=action, state=state, qvalue=q)
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        batch.to_torch()
+
+        non_final_mask = torch.tensor(
+            tuple(map(lambda s: not s, batch.done)),
+            device=self._device,
+            dtype=torch.bool,
+        )
```

Comment:
```suggestion
        non_final_mask = ~batch.done
```
Please use vectorized operation because that is cleaner and faster.